src/main/java/com/google/devtools/build/lib/exec/SpawnLogContext.java

```diff
@@ -119,6 +120,16 @@ protected Platform getPlatform(Spawn spawn, RemoteOptions remoteOptions)
     return builder.build();
   }
 
+  /**
+   * Determines whether an action input is a directory, avoiding I/O if possible.
+   *
+   * <p>Do not call for action outputs.
+   */
+  protected boolean isInputDirectory(Artifact input, InputMetadataProvider inputMetadataProvider)
+      throws IOException {
+    return isInputDirectory(input, input.getPath(), inputMetadataProvider);
```

Comment:
As I recall, the explicit `path` argument is required to make sure that the path was obtained from the action filesystem (usually a `RemoteActionFileSystem`), as otherwise `input.getPath()` will attempt to read from disk, which would fail when building without the bytes. Can you please verify that this is not an issue? (Note that `Path.isDirectory()` returns `false` for a nonexistent path and the vast majority of calls to `isInputDirectory` manage to avoid I/O, so the failure may be rare and silent.)

Reply:
To be more specific, I think the only scenario where this matters is when an output declared as a file is materialized as a directory. For source directories, `isInputDirectory` should be able to exit early at the metadata check; and for file outputs materialized as files, `Path.isDirectory()` returns the correct result when the file is missing from disk.

My plan is to hardcode `--incompatible_disallow_unsound_directory_outputs` in Bazel 8 (which would render this case moot) but I'd still prefer this change to be submitted separately, in case I'm wrong and the `Path.isDirectory()` check turns out to matter in some other case (we've had a spatter of annoying bugs around this).

Reply:
Isn't this covered by `testFileOutputWithDirectoryContents`? I changed the logic to always go through `fileSystem` just in case.

Reply:
Yes, it should have been covered. It's possible that it's no longer needed. I'll try to simplify in a followup; I just prefer not to risk a rollback of this very large PR.